src/puppetlabs/puppetdb/middleware.clj

```diff
@@ -47,36 +48,78 @@
    :post [(fn? %)]}
   (let [allowed? (kitchensink/cn-whitelist->authorizer whitelist)]
     (fn [{:keys [ssl-client-cn] :as req}]
-      (if (allowed? req)
-        :authorized
-        (do
-          (log/warnf "%s rejected by certificate whitelist %s" ssl-client-cn whitelist)
-          (format (str "The client certificate name (%s) doesn't "
-                       "appear in the certificate whitelist. Is your "
-                       "master's (or other PuppetDB client's) certname "
-                       "listed in PuppetDB's certificate-whitelist file?")
-                  ssl-client-cn))))))
+      (when-not (allowed? req)
+        (log/warnf "%s rejected by certificate whitelist %s" ssl-client-cn whitelist)
+        (http/json-response {:error (format (str "Permission denied: The client "
```

Comment:
Switching to json error messages is an API breaking change. There are ways and means for us to start utilizing JSON error messages, but we'd need some way to allow users to switch them on (thus deciding to break API on their own) or doing it in a v5.